# Kenya Agriculture Data Analysis using PySpark

## Data Loading

This notebook demonstrates the data loading process for the Kenya Agriculture dataset using Apache Spark.

The objectives are to:

- Create a Spark session
- Load the dataset
- Inspect the data
- Verify the schema
- Check data quality before preprocessing

In [28]:
%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [29]:
import os
import sys
import importlib
import scripts.load_data
sys.path.append(os.path.abspath(".."))

importlib.reload(scripts.load_data)

<module 'scripts.load_data' from 'c:\\Users\\ADMIN\\OneDrive\\Desktop\\OUK Msc\\1.2\\Data Mining and Big Data\\Big-Data-Analytics-with-Hadoop-and-Spark\\scripts\\load_data.py'>

In [7]:




from scripts.load_data import (
    create_spark_session,
    load_dataset,
    preview_data,
    dataset_shape,
    dataset_schema,
    list_columns,
    dataset_summary,
    check_missing_values,
)

In [8]:
spark = create_spark_session()

In [9]:
# Load the dataset 

df = load_dataset(
    spark,
    "../data/kenya_agriculture_raw.csv"
)

Dataset loaded successfully.


## Preview Dataset

The first five rows are displayed to verify that the dataset has been loaded correctly and to understand its overall structure.

In [10]:
preview_data(df)

+----+---------+----------+---------+---------------+-----+---------+---------------+--------------------+------------+--------------+---------+----+----+-----+----+----+
|Iso3|StartDate|   EndDate|Area Code|Area Code (M49)| Area|Item Code|Item Code (CPC)|                Item|Element Code|       Element|Year Code|Year|Unit|Value|Flag|Note|
+----+---------+----------+---------+---------------+-----+---------+---------------+--------------------+------------+--------------+---------+----+----+-----+----+----+
| KEN| 1/1/1991|12/31/1991|      114|           '404|Kenya|      809|      '01929.07|Abaca, manila hem...|        5312|Area harvested|     1991|1991|  ha| 25.0|   E|NULL|
| KEN| 1/1/1992|12/31/1992|      114|           '404|Kenya|      809|      '01929.07|Abaca, manila hem...|        5312|Area harvested|     1992|1992|  ha| 26.0|   E|NULL|
| KEN| 1/1/1993|12/31/1993|      114|           '404|Kenya|      809|      '01929.07|Abaca, manila hem...|        5312|Area harvested|     1993|1

## Dataset Dimensions

The number of rows and columns provides an overview of the dataset size.

In [11]:
dataset_shape(df)

(24233, 17)

## Dataset Schema

The schema shows the name and data type of each column. Spark automatically inferred the appropriate data types during loading.

In [12]:
dataset_schema(df)

root
 |-- Iso3: string (nullable = true)
 |-- StartDate: string (nullable = true)
 |-- EndDate: string (nullable = true)
 |-- Area Code: integer (nullable = true)
 |-- Area Code (M49): string (nullable = true)
 |-- Area: string (nullable = true)
 |-- Item Code: integer (nullable = true)
 |-- Item Code (CPC): string (nullable = true)
 |-- Item: string (nullable = true)
 |-- Element Code: integer (nullable = true)
 |-- Element: string (nullable = true)
 |-- Year Code: integer (nullable = true)
 |-- Year: integer (nullable = true)
 |-- Unit: string (nullable = true)
 |-- Value: double (nullable = true)
 |-- Flag: string (nullable = true)
 |-- Note: string (nullable = true)



## Column Names

Listing the column names helps verify that all expected variables have been loaded.

In [13]:
print(list_columns(df))

['Iso3', 'StartDate', 'EndDate', 'Area Code', 'Area Code (M49)', 'Area', 'Item Code', 'Item Code (CPC)', 'Item', 'Element Code', 'Element', 'Year Code', 'Year', 'Unit', 'Value', 'Flag', 'Note']


## Summary Statistics

Descriptive statistics provide an overview of the numeric variables, including the count, mean, standard deviation, minimum, and maximum values.

In [14]:
dataset_summary(df)

+-------+-----+---------+----------+---------+---------------+-----+------------------+---------------+--------------------+-----------------+--------------------+------------------+------------------+-------+------------------+-----+-----------------+
|summary| Iso3|StartDate|   EndDate|Area Code|Area Code (M49)| Area|         Item Code|Item Code (CPC)|                Item|     Element Code|             Element|         Year Code|              Year|   Unit|             Value| Flag|             Note|
+-------+-----+---------+----------+---------+---------------+-----+------------------+---------------+--------------------+-----------------+--------------------+------------------+------------------+-------+------------------+-----+-----------------+
|  count|24233|    24233|     24233|    24233|          24233|24233|             24233|          24233|               24233|            24233|               24233|             24233|             24233|  24233|             23502|24233|       

## Missing Values

The dataset is checked for missing values to ensure completeness before any preprocessing or analysis.

In [15]:
check_missing_values(df)

+----+---------+-------+---------+---------------+----+---------+---------------+----+------------+-------+---------+----+----+-----+----+-----+
|Iso3|StartDate|EndDate|Area Code|Area Code (M49)|Area|Item Code|Item Code (CPC)|Item|Element Code|Element|Year Code|Year|Unit|Value|Flag| Note|
+----+---------+-------+---------+---------------+----+---------+---------------+----+------------+-------+---------+----+----+-----+----+-----+
|   0|        0|      0|        0|              0|   0|        0|              0|   0|           0|      0|        0|   0|   0|  731|   0|23708|
+----+---------+-------+---------+---------------+----+---------+---------------+----+------------+-------+---------+----+----+-----+----+-----+



# Data Cleaning

This section evaluates and improves the quality of the dataset by checking for duplicate records, missing values, standardized column names, and invalid numeric values.

In [16]:
import importlib
import scripts.clean_data

importlib.reload(scripts.clean_data)

<module 'scripts.clean_data' from 'c:\\Users\\ADMIN\\OneDrive\\Desktop\\OUK Msc\\1.2\\Data Mining and Big Data\\Big-Data-Analytics-with-Hadoop-and-Spark\\scripts\\clean_data.py'>

In [17]:
from scripts.clean_data import (
    check_duplicates,
    remove_duplicates,
    check_missing_values,
    remove_missing_values,
    standardize_column_names,
    validate_numeric_columns,
    data_quality_report,
)

In [18]:
# Number of duplicate rows
duplicates = check_duplicates(df)
print(f"Duplicate Rows: {duplicates}")

Duplicate Rows: 0


In [19]:
# Missing values
check_missing_values(df)

+----+---------+-------+---------+---------------+----+---------+---------------+----+------------+-------+---------+----+----+-----+----+-----+
|Iso3|StartDate|EndDate|Area Code|Area Code (M49)|Area|Item Code|Item Code (CPC)|Item|Element Code|Element|Year Code|Year|Unit|Value|Flag| Note|
+----+---------+-------+---------+---------------+----+---------+---------------+----+------------+-------+---------+----+----+-----+----+-----+
|   0|        0|      0|        0|              0|   0|        0|              0|   0|           0|      0|        0|   0|   0|  731|   0|23708|
+----+---------+-------+---------+---------------+----+---------+---------------+----+------------+-------+---------+----+----+-----+----+-----+



In [ ]:
df.filter(col("Value").isNull()).show(10, truncate=False)

+----+---------+----------+---------+---------------+-----+---------+---------------+-------------------------------------------------------------------------+------------+--------------+---------+----+----+-----+----+----+
|Iso3|StartDate|EndDate   |Area Code|Area Code (M49)|Area |Item Code|Item Code (CPC)|Item                                                                     |Element Code|Element       |Year Code|Year|Unit|Value|Flag|Note|
+----+---------+----------+---------+---------------+-----+---------+---------------+-------------------------------------------------------------------------+------------+--------------+---------+----+----+-----+----+----+
|KEN |1/1/1961 |12/31/1961|114      |'404           |Kenya|711      |'01654         |Anise, badian, coriander, cumin, caraway, fennel and juniper berries, raw|5312        |Area harvested|1961     |1961|ha  |NULL |M   |NULL|
|KEN |1/1/1962 |12/31/1962|114      |'404           |Kenya|711      |'01654         |Anise, badian, cori

### Missing Values Summary

The missing value assessment indicates that most variables are complete.

Two columns contain missing values:

- **Value** contains 731 missing observations. Since this variable represents the agricultural measurement, these records require further investigation before analysis.
- **Note** contains 23,708 missing values. This column stores optional annotations and therefore its missing values are expected and do not necessarily affect data quality.

Overall, the dataset is largely complete, with missing values concentrated in a small number of fields.

In [20]:
from scripts.clean_data import remove_missing_values
clean_df = remove_missing_values(df)

print("Rows before cleaning:", df.count())
print("Rows after cleaning:", clean_df.count())

Rows before cleaning: 24233
Rows after cleaning: 23502


### Handling Missing Values

The missing value assessment showed that the `Value` column contains 731 missing observations. Inspection revealed that these records are marked with the flag `M`, indicating that the measurements were unavailable rather than equal to zero.

To preserve the integrity of the dataset, missing values were not imputed. Instead, records with missing `Value` entries were excluded from the cleaned dataset used for subsequent transformations and analysis.

The `Note` column contains many missing values because it stores optional annotations. These missing values do not affect the analytical variables and were therefore retained.

In [23]:
# Standardize column names

from scripts.clean_data import standardize_column_names
clean_df = standardize_column_names(clean_df)
print(clean_df.columns)

['iso3', 'startdate', 'enddate', 'area_code', 'area_code_m49', 'area', 'item_code', 'item_code_cpc', 'item', 'element_code', 'element', 'year_code', 'year', 'unit', 'value', 'flag', 'note']


In [31]:
from pyspark.sql.functions import col
clean_df.filter(col("value") < 0).count()

0

In [32]:
# Validate numeric values
invalid = validate_numeric_columns(clean_df)
invalid.show(5)

Negative values found: 0
+----+---------+-------+---------+-------------+----+---------+-------------+----+------------+-------+---------+----+----+-----+----+----+
|iso3|startdate|enddate|area_code|area_code_m49|area|item_code|item_code_cpc|item|element_code|element|year_code|year|unit|value|flag|note|
+----+---------+-------+---------+-------------+----+---------+-------------+----+------------+-------+---------+----+----+-----+----+----+
+----+---------+-------+---------+-------------+----+---------+-------------+----+------------+-------+---------+----+----+-----+----+----+



In [33]:
# Data quality report
data_quality_report(clean_df)

Rows: 23502
Columns: 17
Duplicates: 0
Missing Values
+----+---------+-------+---------+-------------+----+---------+-------------+----+------------+-------+---------+----+----+-----+----+-----+
|iso3|startdate|enddate|area_code|area_code_m49|area|item_code|item_code_cpc|item|element_code|element|year_code|year|unit|value|flag| note|
+----+---------+-------+---------+-------------+----+---------+-------------+----+------------+-------+---------+----+----+-----+----+-----+
|   0|        0|      0|        0|            0|   0|        0|            0|   0|           0|      0|        0|   0|   0|    0|   0|22977|
+----+---------+-------+---------+-------------+----+---------+-------------+----+------------+-------+---------+----+----+-----+----+-----+



## Data Cleaning Summary

The dataset was successfully cleaned and validated before further analysis.

The cleaning process included:

- Checking for duplicate records and confirming that none remained.
- Removing records with missing values in the `value` column, as these represent unavailable agricultural measurements.
- Standardizing column names by converting them to lowercase and replacing spaces with underscores.
- Validating the dataset to ensure all analytical variables were complete.

Following the cleaning process, the dataset contains **23,502 records** and **17 variables**. The only remaining missing values occur in the `note` column, which contains optional metadata and does not affect the analytical variables used in this project.

# Data Transformations 
- filter_by_element() ensures you're analyzing one type of agricultural measure (such as Production or Yield) at a time.
- select_analysis_columns() removes unnecessary fields, making processing more efficient.
- aggregate_yearly_production() reveals production trends over time.
- aggregate_crop_production() identifies the crops with the highest total production.
- aggregate_crop_year() supports detailed trend analysis for each crop across years.
- sort_by_year() ensures chronological ordering for reporting and visualization.

In [40]:
from scripts.transformations import (
    filter_by_element,
    select_analysis_columns,
    aggregate_yearly_production,
    prepare_analysis_dataset,
    aggregate_crop_production,
    aggregate_crop_year,
    sort_by_year,
)

In [48]:
# Filter  data

# Production
production_df = filter_by_element(clean_df, "Production")
production_df.show(3)

# Area Harvested
harvest_df = filter_by_element(clean_df, "Area harvested")
harvest_df.show(3)

# Yield
yield_df = filter_by_element(clean_df, "Yield")
yield_df.show(3)

#Stocks 

stocks_df = filter_by_element(clean_df, "Stocks")
stocks_df.show(3)

#laying 
laying_df = filter_by_element(clean_df, "Laying")
laying_df.show(3)

#Milk anaimals

milk_animals_df = filter_by_element(clean_df, "Milk Animals")
milk_animals_df.show(3)

#"Producing Animals/Slaughtered 

slaughtered_df = filter_by_element(
    clean_df,
    "Producing Animals/Slaughtered"
)
slaughtered_df.show(3)

#car"Yield/Carcass Weight

carcass_df = filter_by_element(
    clean_df,
    "Yield/Carcass Weight"
)
carcass_df.show(3)


+----+---------+----------+---------+-------------+-----+---------+-------------+--------------------+------------+----------+---------+----+----+-----+----+----+
|iso3|startdate|   enddate|area_code|area_code_m49| area|item_code|item_code_cpc|                item|element_code|   element|year_code|year|unit|value|flag|note|
+----+---------+----------+---------+-------------+-----+---------+-------------+--------------------+------------+----------+---------+----+----+-----+----+----+
| KEN| 1/1/1961|12/31/1961|      114|         '404|Kenya|      809|    '01929.07|Abaca, manila hem...|        5510|Production|     1961|1961|   t|  0.0|   A|NULL|
| KEN| 1/1/1962|12/31/1962|      114|         '404|Kenya|      809|    '01929.07|Abaca, manila hem...|        5510|Production|     1962|1962|   t|  0.0|   A|NULL|
| KEN| 1/1/1963|12/31/1963|      114|         '404|Kenya|      809|    '01929.07|Abaca, manila hem...|        5510|Production|     1963|1963|   t|  0.0|   A|NULL|
+----+---------+------

In [49]:
# Production
production_analysis_df = prepare_analysis_dataset(production_df)

# Area Harvested
harvest_analysis_df = prepare_analysis_dataset(harvest_df)

# Yield
yield_analysis_df = prepare_analysis_dataset(yield_df)

# Stocks
stocks_analysis_df = prepare_analysis_dataset(stocks_df)

# Laying
laying_analysis_df = prepare_analysis_dataset(laying_df)

# Milk Animals
milk_analysis_df = prepare_analysis_dataset(milk_animals_df)

# Producing Animals / Slaughtered
slaughtered_analysis_df = prepare_analysis_dataset(slaughtered_df)

# Yield / Carcass Weight
carcass_analysis_df = prepare_analysis_dataset(carcass_df)

In [51]:
datasets = {
    "Production": production_analysis_df,
    "Area Harvested": harvest_analysis_df,
    "Yield": yield_analysis_df,
    "Stocks": stocks_analysis_df,
    "Laying": laying_analysis_df,
    "Milk Animals": milk_analysis_df,
    "Producing Animals/Slaughtered": slaughtered_analysis_df,
    "Yield/Carcass Weight": carcass_analysis_df
}

for name, df in datasets.items():
    print(f"\n{name}")
    print(f"Rows: {df.count()}")
    print(f"Columns: {len(df.columns)}")
    df.show(3)


Production
Rows: 9278
Columns: 5
+--------------------+----------+----+----+-----+
|                item|   element|year|unit|value|
+--------------------+----------+----+----+-----+
|Abaca, manila hem...|Production|1961|   t|  0.0|
|Abaca, manila hem...|Production|1962|   t|  0.0|
|Abaca, manila hem...|Production|1963|   t|  0.0|
+--------------------+----------+----+----+-----+
only showing top 3 rows


Area Harvested
Rows: 5100
Columns: 5
+--------------------+--------------+----+----+-----+
|                item|       element|year|unit|value|
+--------------------+--------------+----+----+-----+
|Abaca, manila hem...|Area harvested|1991|  ha| 25.0|
|Abaca, manila hem...|Area harvested|1992|  ha| 26.0|
|Abaca, manila hem...|Area harvested|1993|  ha| 26.0|
+--------------------+--------------+----+----+-----+
only showing top 3 rows


Yield
Rows: 5164
Columns: 5
+--------------------+-------+----+-----+------+
|                item|element|year| unit| value|
+--------------------+-

In [50]:
analysis_df = prepare_analysis_dataset(clean_df)

analysis_df.show(5)

+--------------------+--------------+----+----+-----+
|                item|       element|year|unit|value|
+--------------------+--------------+----+----+-----+
|Abaca, manila hem...|Area harvested|1991|  ha| 25.0|
|Abaca, manila hem...|Area harvested|1992|  ha| 26.0|
|Abaca, manila hem...|Area harvested|1993|  ha| 26.0|
|Abaca, manila hem...|Area harvested|1994|  ha| 27.0|
|Abaca, manila hem...|Area harvested|1995|  ha| 28.0|
+--------------------+--------------+----+----+-----+
only showing top 5 rows



In [52]:
#Total production by year

yearly_df = aggregate_yearly_production(analysis_df)

yearly_df.show(10)

+----+-------------------+
|year|   total_production|
+----+-------------------+
|1961|7.284519225999999E7|
|1962|7.308677810000001E7|
|1963|7.303660412999998E7|
|1964|7.702543749999999E7|
|1965|      7.774897056E7|
|1966|      7.752754827E7|
|1967|      7.979995948E7|
|1968|      8.102047744E7|
|1969|       7.94385718E7|
|1970|      8.057674172E7|
+----+-------------------+
only showing top 10 rows



In [ ]:
# top crops by production

crop_df = aggregate_crop_production(analysis_df)

crop_df.show(10)

+--------------------+--------------------+
|                item|    total_production|
+--------------------+--------------------+
| Sugar Crops Primary|      2.5909036424E8|
|          Sugar cane|      2.5909036424E8|
|    Cereals, primary|1.9741207748999998E8|
|         Milk, Total|1.7972032641999996E8|
|        Maize (corn)|      1.5881388144E8|
|  Raw milk of cattle|       1.441147808E8|
|Roots and Tubers,...|1.3598624954000002E8|
|       Fruit Primary|1.1907329065999997E8|
|  Vegetables Primary| 8.246536838999999E7|
|Hen eggs in shell...|        6.72914964E7|
+--------------------+--------------------+
only showing top 10 rows



In [ ]:
# Production by crop and year

crop_year_df = aggregate_crop_year(analysis_df)

crop_year_df.show(10)

+--------------------+----+----------------+
|                item|year|total_production|
+--------------------+----+----------------+
|Abaca, manila hem...|1961|             0.0|
|Abaca, manila hem...|1962|             0.0|
|Abaca, manila hem...|1963|             0.0|
|Abaca, manila hem...|1964|             0.0|
|Abaca, manila hem...|1965|             0.0|
|Abaca, manila hem...|1966|             0.0|
|Abaca, manila hem...|1967|             0.0|
|Abaca, manila hem...|1968|             0.0|
|Abaca, manila hem...|1969|             0.0|
|Abaca, manila hem...|1970|             0.0|
+--------------------+----+----------------+
only showing top 10 rows



# Exploratory Data Analysis (EDA)

The purpose of Exploratory Data Analysis (EDA) is to summarize the main characteristics of the transformed datasets and identify patterns, trends, and anomalies before answering the research questions.

The analysis is performed separately for each agricultural element to provide insights into:

- Dataset size
- Summary statistics
- Top and lowest performing agricultural items
- Trends over time
- Measurement units
- Descriptive statistics

In [56]:
from scripts.analysis import (
    dataset_overview,
    summary_statistics,
    top_items,
    lowest_items,
    values_by_year,
    highest_year,
    lowest_year,
    average_value,
    unique_items,
    units,
    value_statistics,
    item_statistics,
    year_statistics,
    element_statistics
)

In [58]:
#Dataset Overview

dataset_overview(production_analysis_df)

dataset_overview(harvest_analysis_df)

dataset_overview(yield_analysis_df)

dataset_overview(stocks_analysis_df)

dataset_overview(laying_analysis_df)

dataset_overview(milk_analysis_df)

dataset_overview(slaughtered_analysis_df)

dataset_overview(carcass_analysis_df)

DATASET OVERVIEW
Rows: 9278
Columns: 5
Columns: ['item', 'element', 'year', 'unit', 'value']
DATASET OVERVIEW
Rows: 5100
Columns: 5
Columns: ['item', 'element', 'year', 'unit', 'value']
DATASET OVERVIEW
Rows: 5164
Columns: 5
Columns: ['item', 'element', 'year', 'unit', 'value']
DATASET OVERVIEW
Rows: 768
Columns: 5
Columns: ['item', 'element', 'year', 'unit', 'value']
DATASET OVERVIEW
Rows: 128
Columns: 5
Columns: ['item', 'element', 'year', 'unit', 'value']
DATASET OVERVIEW
Rows: 320
Columns: 5
Columns: ['item', 'element', 'year', 'unit', 'value']
DATASET OVERVIEW
Rows: 1472
Columns: 5
Columns: ['item', 'element', 'year', 'unit', 'value']
DATASET OVERVIEW
Rows: 1272
Columns: 5
Columns: ['item', 'element', 'year', 'unit', 'value']


In [59]:
# Summary statisticts

summary_statistics(production_analysis_df).show()

summary_statistics(harvest_analysis_df).show()

summary_statistics(yield_analysis_df).show()

summary_statistics(stocks_analysis_df).show()

summary_statistics(laying_analysis_df).show()

summary_statistics(milk_analysis_df).show()

summary_statistics(slaughtered_analysis_df).show()

summary_statistics(carcass_analysis_df).show()

+-------+------------------+
|summary|             value|
+-------+------------------+
|  count|              9278|
|   mean|253978.05763203368|
| stddev| 785014.6967334384|
|    min|               0.0|
|    max|         9365300.0|
+-------+------------------+

+-------+------------------+
|summary|             value|
+-------+------------------+
|  count|              5100|
|   mean|106449.59490196078|
| stddev| 330295.7039852331|
|    min|               1.0|
|    max|         2940254.0|
+-------+------------------+

+-------+------------------+
|summary|             value|
+-------+------------------+
|  count|              5164|
|   mean|7511.1288729666885|
| stddev| 13580.47657089705|
|    min|              33.3|
|    max|          121184.5|
+-------+------------------+

+-------+--------------------+
|summary|               value|
+-------+--------------------+
|  count|                 768|
|   mean|   6491712.571614583|
| stddev|1.0004176983802302E7|
|    min|                 0.

In [60]:
#Top Items

top_items(production_analysis_df).show()

top_items(harvest_analysis_df).show()

top_items(yield_analysis_df).show()

top_items(stocks_analysis_df).show()

top_items(laying_analysis_df).show()

top_items(milk_analysis_df).show()

top_items(slaughtered_analysis_df).show()

top_items(carcass_analysis_df).show()

+--------------------+--------------------+
|                item|         total_value|
+--------------------+--------------------+
| Sugar Crops Primary|      2.5909036424E8|
|          Sugar cane|      2.5909036424E8|
|    Cereals, primary|1.9741207748999998E8|
|         Milk, Total|1.7972032641999996E8|
|        Maize (corn)|      1.5881388144E8|
|  Raw milk of cattle|       1.441147808E8|
|Roots and Tubers,...|1.3598624954000002E8|
|       Fruit Primary|1.1907329065999997E8|
|  Vegetables Primary| 8.246536838999999E7|
|Hen eggs in shell...|        6.72914964E7|
+--------------------+--------------------+

+--------------------+------------+
|                item| total_value|
+--------------------+------------+
|    Cereals, primary|1.28648116E8|
|        Maize (corn)| 1.0085278E8|
|       Pulses, Total| 7.7073263E7|
|          Beans, dry| 4.0920596E7|
| Other pulses n.e.c.|  2.358427E7|
|Roots and Tubers,...| 1.3134836E7|
|Oilcrops, Oil Equ...| 1.2128956E7|
|Oilcrops, Cake Eq...| 

In [61]:
#Lowest Items

lowest_items(production_analysis_df).show()

lowest_items(harvest_analysis_df).show()

lowest_items(yield_analysis_df).show()

lowest_items(stocks_analysis_df).show()

lowest_items(laying_analysis_df).show()

lowest_items(milk_analysis_df).show()

lowest_items(slaughtered_analysis_df).show()

lowest_items(carcass_analysis_df).show()

+--------------------+------------------+
|                item|       total_value|
+--------------------+------------------+
|           Whey, dry|              84.0|
|          Areca nuts|            504.76|
|        Vanilla, raw|            523.26|
|          Artichokes|1072.7699999999998|
|Abaca, manila hem...|1771.6099999999994|
|Nutmeg, mace, car...|2047.9099999999999|
|            Apricots|           2283.63|
|Whole milk, evapo...| 5542.169999999999|
|           Asparagus| 7592.160000000001|
|         Ginger, raw|           10813.6|
+--------------------+------------------+

+--------------------+-----------+
|                item|total_value|
+--------------------+-----------+
|          Artichokes|      171.0|
|            Apricots|      442.0|
|           Asparagus|     1179.0|
|Abaca, manila hem...|     1200.0|
|        Vanilla, raw|     1622.0|
|Peaches and necta...|     1649.0|
|         Ginger, raw|     1713.0|
|        Strawberries|     2101.0|
|Leeks and other a...|    

In [63]:
#Value over the years 
values_by_year(production_analysis_df).show(5)

values_by_year(harvest_analysis_df).show(5)

values_by_year(yield_analysis_df).show(5)

values_by_year(stocks_analysis_df).show(5)

values_by_year(laying_analysis_df).show(5)

values_by_year(milk_analysis_df).show(5)

values_by_year(slaughtered_analysis_df).show(5)

values_by_year(carcass_analysis_df).show(5)

+----+--------------------+
|year|         total_value|
+----+--------------------+
|1961|       1.031794096E7|
|1962|1.0879274299999999E7|
|1963|       1.132073823E7|
|1964|1.1875766800000003E7|
|1965|       1.180776526E7|
+----+--------------------+
only showing top 5 rows

+----+-----------+
|year|total_value|
+----+-----------+
|1961|  4471883.0|
|1962|  4903107.0|
|1963|  5133505.0|
|1964|  5316862.0|
|1965|  5618429.0|
+----+-----------+
only showing top 5 rows

+----+------------------+
|year|       total_value|
+----+------------------+
|1961|          282818.3|
|1962|289107.80000000005|
|1963|293386.89999999997|
|1964| 395894.6999999999|
|1965|302383.30000000005|
+----+------------------+
only showing top 5 rows

+----+-----------+
|year|total_value|
+----+-----------+
|1961|  3.33281E7|
|1962|  3.32133E7|
|1963|  3.25857E7|
|1964|   3.4564E7|
|1965|  3.47752E7|
+----+-----------+
only showing top 5 rows

+----+-----------+
|year|total_value|
+----+-----------+
|1961|     8000

In [64]:
#best year for each dataset

highest_year(production_analysis_df).show()

highest_year(harvest_analysis_df).show()

highest_year(yield_analysis_df).show()

highest_year(stocks_analysis_df).show()

highest_year(laying_analysis_df).show()

highest_year(milk_analysis_df).show()

highest_year(slaughtered_analysis_df).show()

highest_year(carcass_analysis_df).show()

+----+-------------------+
|year|        total_value|
+----+-------------------+
|2024|7.600948392999998E7|
+----+-------------------+

+----+-----------+
|year|total_value|
+----+-----------+
|2023|1.2946148E7|
+----+-----------+

+----+-----------------+
|year|      total_value|
+----+-----------------+
|2008|909178.8999999998|
+----+-----------------+

+----+------------+
|year| total_value|
+----+------------+
|2020|1.83128719E8|
+----+------------+

+----+-----------+
|year|total_value|
+----+-----------+
|2021|    42840.0|
+----+-----------+

+----+-----------+
|year|total_value|
+----+-----------+
|2024|3.0703841E7|
+----+-----------+

+----+-----------+
|year|total_value|
+----+-----------+
|2019|7.4021541E7|
+----+-----------+

+----+-----------+
|year|total_value|
+----+-----------+
|2010|    17471.0|
+----+-----------+



In [65]:
#lowest year for each dataset
lowest_year(production_analysis_df).show()

lowest_year(harvest_analysis_df).show()

lowest_year(yield_analysis_df).show()

lowest_year(stocks_analysis_df).show()

lowest_year(laying_analysis_df).show()

lowest_year(milk_analysis_df).show()

lowest_year(slaughtered_analysis_df).show()

lowest_year(carcass_analysis_df).show()

+----+-------------+
|year|  total_value|
+----+-------------+
|1961|1.031794096E7|
+----+-------------+

+----+-----------+
|year|total_value|
+----+-----------+
|1961|  4471883.0|
+----+-----------+

+----+-----------+
|year|total_value|
+----+-----------+
|1961|   282818.3|
+----+-----------+

+----+-----------+
|year|total_value|
+----+-----------+
|1963|  3.25857E7|
+----+-----------+

+----+-----------+
|year|total_value|
+----+-----------+
|1966|     7700.0|
+----+-----------+

+----+-----------+
|year|total_value|
+----+-----------+
|1961|  6632000.0|
+----+-----------+

+----+-----------+
|year|total_value|
+----+-----------+
|1973|1.5666065E7|
+----+-----------+

+----+-----------+
|year|total_value|
+----+-----------+
|1961|     8630.0|
+----+-----------+



In [66]:
#value statistics

value_statistics(production_analysis_df).show()

value_statistics(harvest_analysis_df).show()

value_statistics(yield_analysis_df).show()

value_statistics(stocks_analysis_df).show()

value_statistics(laying_analysis_df).show()

value_statistics(milk_analysis_df).show()

value_statistics(slaughtered_analysis_df).show()

value_statistics(carcass_analysis_df).show()

+-------------+-------------+------------------+
|minimum_value|maximum_value|     average_value|
+-------------+-------------+------------------+
|          0.0|    9365300.0|253978.05763203368|
+-------------+-------------+------------------+

+-------------+-------------+------------------+
|minimum_value|maximum_value|     average_value|
+-------------+-------------+------------------+
|          1.0|    2940254.0|106449.59490196078|
+-------------+-------------+------------------+

+-------------+-------------+------------------+
|minimum_value|maximum_value|     average_value|
+-------------+-------------+------------------+
|         33.3|     121184.5|7511.1288729666885|
+-------------+-------------+------------------+

+-------------+-------------+-----------------+
|minimum_value|maximum_value|    average_value|
+-------------+-------------+-----------------+
|          0.0|  6.4641088E7|6491712.571614583|
+-------------+-------------+-----------------+

+-------------+------

In [69]:
#Item statistics

item_statistics(production_analysis_df).show()

item_statistics(harvest_analysis_df).show()

item_statistics(yield_analysis_df).show()

item_statistics(stocks_analysis_df).show()

item_statistics(laying_analysis_df).show()

item_statistics(milk_analysis_df).show()

item_statistics(slaughtered_analysis_df).show()

item_statistics(carcass_analysis_df).show()

+---------------+
|number_of_items|
+---------------+
|            161|
+---------------+

+---------------+
|number_of_items|
+---------------+
|             96|
+---------------+

+---------------+
|number_of_items|
+---------------+
|             97|
+---------------+

+---------------+
|number_of_items|
+---------------+
|             12|
+---------------+

+---------------+
|number_of_items|
+---------------+
|              2|
+---------------+

+---------------+
|number_of_items|
+---------------+
|              5|
+---------------+

+---------------+
|number_of_items|
+---------------+
|             23|
+---------------+

+---------------+
|number_of_items|
+---------------+
|             20|
+---------------+





# This project aims to analyze Kenya's agricultural production data to uncover trends, identify high- and low-performing crops, and generate insights that can support data-driven agricultural decision-making.



### 1. How has agricultural production changed over time?

This analysis examines yearly agricultural production in Kenya to determine whether production has increased, decreased, or remained stable over time.

In [71]:
values_by_year(production_analysis_df).show()

+----+--------------------+
|year|         total_value|
+----+--------------------+
|1961|       1.031794096E7|
|1962|1.0879274299999999E7|
|1963|       1.132073823E7|
|1964|1.1875766800000003E7|
|1965|       1.180776526E7|
|1966|1.1718647270000001E7|
|1967|1.2894985880000003E7|
|1968|       1.437415904E7|
|1969|        1.54546259E7|
|1970|       1.585350512E7|
|1971|       1.583133439E7|
|1972|       1.600012258E7|
|1973|       1.783699886E7|
|1974|       1.859268438E7|
|1975|1.8704301810000002E7|
|1976|       2.018617912E7|
|1977|       2.117223539E7|
|1978|       2.187972067E7|
|1979|2.4351764840000004E7|
|1980|       2.522870558E7|
+----+--------------------+
only showing top 20 rows



### 2. Which crops have the highest production?

The following table identifies the crops with the highest total production recorded in the dataset.

In [72]:
top_items(production_analysis_df).show()

+--------------------+--------------------+
|                item|         total_value|
+--------------------+--------------------+
| Sugar Crops Primary|      2.5909036424E8|
|          Sugar cane|      2.5909036424E8|
|    Cereals, primary|1.9741207748999998E8|
|         Milk, Total|1.7972032641999996E8|
|        Maize (corn)|      1.5881388144E8|
|  Raw milk of cattle|       1.441147808E8|
|Roots and Tubers,...|1.3598624954000002E8|
|       Fruit Primary|1.1907329065999997E8|
|  Vegetables Primary| 8.246536838999999E7|
|Hen eggs in shell...|        6.72914964E7|
+--------------------+--------------------+



### 3. Which crops have the lowest production?

The following table identifies crops with the smallest recorded production values.

In [73]:
lowest_items(production_analysis_df).show()

+--------------------+------------------+
|                item|       total_value|
+--------------------+------------------+
|           Whey, dry|              84.0|
|          Areca nuts|            504.76|
|        Vanilla, raw|            523.26|
|          Artichokes|1072.7699999999998|
|Abaca, manila hem...|1771.6099999999994|
|Nutmeg, mace, car...|2047.9099999999999|
|            Apricots|           2283.63|
|Whole milk, evapo...| 5542.169999999999|
|           Asparagus| 7592.160000000001|
|         Ginger, raw|           10813.6|
+--------------------+------------------+



### 4. Which year recorded the highest agricultural production?

The following result identifies the year with the greatest total agricultural production.

In [74]:
highest_year(production_analysis_df).show()

+----+-------------------+
|year|        total_value|
+----+-------------------+
|2024|7.600948392999998E7|
+----+-------------------+



### 5. Which year recorded the lowest agricultural production?

The following result identifies the year with the lowest agricultural production.

In [75]:
lowest_year(production_analysis_df).show()

+----+-------------+
|year|  total_value|
+----+-------------+
|1961|1.031794096E7|
+----+-------------+



### 6. What is the average agricultural production?

The average production value provides an overall estimate of agricultural output across the dataset.

In [76]:
average_value(production_analysis_df).show()

+------------------+
|     average_value|
+------------------+
|253978.05763203368|
+------------------+



### 7. How many different crops are represented?

This analysis counts the number of unique agricultural commodities contained in the production dataset.

In [77]:
unique_items(production_analysis_df).show()

+------------+
|unique_items|
+------------+
|         161|
+------------+



### 8. Which production units are used?

This analysis identifies the units used to record production values.

In [78]:
units(production_analysis_df).show()

+-------+
|   unit|
+-------+
|1000 No|
|      t|
+-------+

